octoprint_PrintTimeGenius/static/js/PrintTimeGenius.js

```diff
@@ -70,19 +70,25 @@ $(function() {
         }
       }
       return results;
-    }
+    };
 
-    self.FileList = ko.pureComputed(function() {
-      return self.recurseFiles(self.filesViewModel.allItems()).slice()
-          .sort(function(a,b) {
-            if (_.has(a, "gcodeAnalysis.progress") !=
-                _.has(b, "gcodeAnalysis.progress")) {
-              return (_.has(a, "gcodeAnalysis.progress") -
-                      _.has(b, "gcodeAnalysis.progress"));
-            }
-            return a.path.localeCompare(b.path);
-          });
-    });
+    self.theFiles = function(items) {
+    	let results = [];
+    	let queue = [{children: items}];
+
+    	while (queue.length > 0) {
+    		item = queue.shift();
+    		results.push.apply(results, item.children.filter(item => item["type"] == "machinecode"));
+    		queue.push.apply(queue, item.children.filter(item => "children" in item));
+    	}
+    	return results;
+    };
+
+    self.FileList = ko.observableArray([]);
+
+    self.onSettingsShown = function() {
+        self.FileList(self.theFiles(self.filesViewModel.allItems()));
```

Comment:
I bet that it's not the iterative version that saves so much time but rather this change which computes the file list only when the settings are shown.  Does this work in all cases?  Like, what happens if a file upload completes while the settings window is shown?  It won't recompute, right?  Does it recompute in the original version?

It's kind of a minor point and people will likely figure out to close and re-open the options if this rare event occurs so even if the behavior is changed, the time savings are worth it so I'd still be okay with this change.

Reply:
Yes this is where the bulk of the difference comes in, the difference with the iterative version alone is very small, mostly I just like to avoid blowing up the stack :)

I'm not sure if the original recomputes if a file is uploaded or not, it's not clear to me under what circumstances knockout re-evaluates pure computed observables

It's true that this version only recomputes if the settings window is closed and re-opened. I think I can fix that pretty easily though